WalletWasabi.Backend/Filters/IdempotentActionFilterAttributeImpl.cs

```diff
@@ -1,66 +1,114 @@
-using System;
-using System.Security.Cryptography;
-using System.Text;
-using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.AspNetCore.Mvc.Filters;
 using Microsoft.Extensions.Caching.Memory;
 using Newtonsoft.Json;
+using System;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
 using WalletWasabi.WabiSabi.Models.Serialization;
 
 namespace WalletWasabi.Backend.Filters
 {
 	public class IdempotentActionFilterAttributeImpl : ActionFilterAttribute
 	{
-		private readonly IMemoryCache _cache;
+		private static TimeSpan CacheTimeout { get; } = TimeSpan.FromMinutes(5);
+		private const string ContextCacheKey = "cached-key";
 
 		public IdempotentActionFilterAttributeImpl(IMemoryCache cache)
 			: base()
 		{
-			_cache = cache;
+			ResponseCache = cache;
 		}
 
+		/// <summary>Guards <see cref="ResponseCache"/>.</summary>
+		/// <remarks>Unfortunately, <see cref="CacheExtensions.GetOrCreate{TItem}(IMemoryCache, object, Func{ICacheEntry, TItem})"/> is not atomic.</remarks>
+		/// <seealso href="https://github.com/dotnet/runtime/issues/36499"/>
+		private object ResponseCacheLock { get; } = new();
+
+		/// <remarks>Guarded by <see cref="ResponseCacheLock"/>.</remarks>
+		private IMemoryCache ResponseCache { get; }
+
+		/// <inheritdoc/>
 		public override async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
 		{
-			var request = context.HttpContext.Request;
+			HttpRequest request = context.HttpContext.Request;
 
 			if (context.ModelState.IsValid)
 			{
-				if (context.ActionArguments.TryGetValue("request", out var model))
+				if (!context.ActionArguments.TryGetValue("request", out object? model))
 				{
-					var cacheKey = GetCacheEntryKey(request.Path, model);
+					throw new InvalidOperationException("Control actions marked as Idempotent must receive a 'request' argument.");
+				}
+
+				string cacheKey = GetCacheEntryKey(request.Path, model);
+				TaskCompletionSource<IActionResult>? cachedResponseTcs = null;
+				TaskCompletionSource<IActionResult>? responseTcs = null;
 
-					if (_cache.TryGetValue<ObjectResult>(cacheKey, out var cachedResponse))
+				lock (ResponseCacheLock)
+				{
+					if (!ResponseCache.TryGetValue(cacheKey, out cachedResponseTcs))
 					{
-						context.Result = cachedResponse;
-						context.HttpContext.Items.Remove("cached-key");
-						return;
+						responseTcs = new();
+						ResponseCache.Set(cacheKey, responseTcs, DateTimeOffset.UtcNow.Add(CacheTimeout));
 					}
+				}
 
-					context.HttpContext.Items["cached-key"] = cacheKey;
+				if (cachedResponseTcs is not null)
+				{
+					try
+					{
+						context.Result = await cachedResponseTcs!.Task.WithAwaitCancellationAsync(TimeSpan.FromMinutes(1)).ConfigureAwait(false);
```

Comment:
This is okay but if `CacheTimeout` was created  to replace the `TimeSpan.FromXXXXX` then this could be a constant too.

Reply:
Yes, this should be a constant too. Will change.